texar/torch/modules/tokenizers/pretrained_xlnet_tokenizer.py

```diff
@@ -0,0 +1,228 @@
+            and default values.
+    """
+
+    _MAX_INPUT_SIZE = {
+        'xlnet-base-cased': None,
+        'xlnet-large-cased': None,
+    }
+    _VOCAB_FILE_NAMES = {'vocab_file': 'spiece.model'}
+
+    def __init__(self,
+                 pretrained_model_name: Optional[str] = None,
+                 cache_dir: Optional[str] = None,
+                 hparams=None):
+        super().__init__(hparams=hparams)
+
+        self.__dict__: Dict
+
+        self.config = {
+            'do_lower_case': self.hparams['do_lower_case'],
+            'remove_space': self.hparams['remove_space'],
+            'keep_accents': self.hparams['keep_accents'],
+        }
+
+        self.load_pretrained_config(pretrained_model_name, cache_dir)
+
+        if self.pretrained_model_dir is not None:
+            vocab_file = os.path.join(self.pretrained_model_dir,
+                                      self._VOCAB_FILE_NAMES['vocab_file'])
+            assert pretrained_model_name is not None
+            if self._MAX_INPUT_SIZE.get(pretrained_model_name):
+                self.max_len = self._MAX_INPUT_SIZE[pretrained_model_name]
+        else:
+            vocab_file = self.hparams['vocab_file']
+            if self.hparams.get('max_len'):
+                self.max_len = self.hparams['max_len']
+
+        if not os.path.isfile(vocab_file):
+            raise ValueError("Can't find a vocabulary file at path "
+                             "'{}".format(vocab_file))
+
+        self.do_lower_case = self.hparams["do_lower_case"]
+        self.remove_space = self.hparams["remove_space"]
+        self.keep_accents = self.hparams["keep_accents"]
+        self.vocab_file = vocab_file
+
+        self.sp_model = spm.SentencePieceProcessor()
+        self.sp_model.Load(self.vocab_file)
+
+    def forward(self, inputs, task, **kwargs):
+        if task == 'text-to-token':
+            sample = kwargs.get('sample', False)
+            return self.tokenize(inputs, sample=sample)
+        elif task == 'token-to-text':
+            return self.convert_tokens_to_string(inputs)
+        elif task == 'text-to-id':
+            return self.encode(inputs)
+        elif task == 'id-to-text':
+            skip_special_tokens = kwargs.get('skip_special_tokens', False)
+            clean_up_tokenization_spaces = kwargs.get(
+                'clean_up_tokenization_spaces', True)
+            return self.decode(
+                inputs, skip_special_tokens=skip_special_tokens,
+                clean_up_tokenization_spaces=clean_up_tokenization_spaces)
+        elif task == 'token-to-id':
+            return self.convert_tokens_to_ids(inputs)
+        elif task == 'id-to-token':
+            skip_special_tokens = kwargs.get('skip_special_tokens', False)
+            return self.convert_ids_to_tokens(
+                inputs, skip_special_tokens=skip_special_tokens)
+        else:
+            raise ValueError("Unrecognized job type.")
+
+    def __getstate__(self):
```

Comment:
Is pickling only supported for XLNet tokenizer? I don't see `__getstate__` and `__setstate__` implemented for other classes.

Reply:
All tokenizers can be pickled. The reason we implemented `__getstate__` and `__setstate__` only for `XLNetTokenizer` is that `spm.SentencePieceProcessor()` used in `XLNetTokenizer` is SwigPyObject object which cannot be pickled. Other tokenizers do not have such an issue and can be directly pickled.

Reply:
Let's also document this behavior because I think this is a bit different to other tokenizers. To unpickle this tokenizer you need to make sure the vocabulary file is in place, while I don't think this is necessary for the other tokenizers. We may also try to directly save the contents of the vocabulary file when pickling,